coap/server.go

```diff
@@ -1,30 +1,114 @@
-package adapter
+package coap
 
 import (
-	"log"
+	"fmt"
 	"net"
+	"strings"
+	"time"
 
-	"github.com/dereulenspiegel/coap-mux"
-	"github.com/dustin/go-coap"
+	mux "github.com/dereulenspiegel/coap-mux"
+	manager "github.com/mainflux/mainflux/manager/client"
+
+	gocoap "github.com/dustin/go-coap"
 )
 
-func notFoundHandler(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
-	log.Printf("Got message in notFoundHandler: path=%q: %#v from %v", m.Path(), m, a)
-	if m.IsConfirmable() {
-		return &coap.Message{
-			Type: coap.Acknowledgement,
-			Code: coap.NotFound,
+var auth manager.ManagerClient
+
+func authKey(opt interface{}) (string, error) {
+	if opt == nil {
+		return "", errBadRequest
+	}
+	val, ok := opt.(string)
+	if !ok {
+		return "", errBadRequest
+	}
+	arr := strings.Split(val, "=")
+	if len(arr) != 2 || strings.ToLower(arr[0]) != "key" {
+		return "", errBadOption
+	}
+	return arr[1], nil
+}
+
+// Authorize method is used to authorize request.
+func Authorize(msg *gocoap.Message, res *gocoap.Message, cid string) (publisher string, err error) {
+	// Device Key is passed as Uri-Query parameter, which option ID is 15 (0xf).
+	key, err := authKey(msg.Option(gocoap.URIQuery))
+	if err != nil {
+		switch err {
+		case errBadOption:
+			res.Code = gocoap.BadOption
+		case errBadRequest:
+			res.Code = gocoap.BadRequest
+		}
+		return
+	}
+
+	publisher, err = auth.CanAccess(cid, key)
+	if err != nil {
+		switch err {
+		case manager.ErrServiceUnreachable:
+			res.Code = gocoap.ServiceUnavailable
+		default:
+			res.Code = gocoap.Unauthorized
 		}
 	}
-	return nil
+	return
 }
 
-func (ca *CoAPAdapter) COAPServer() *mux.Router {
-	r := mux.NewRouter()
-	r.Handle("/channels/{channel_id}/messages", coap.FuncHandler(ca.sendMessage)).Methods(coap.POST)
-	r.Handle("/channels/{channel_id}/messages", coap.FuncHandler(ca.observeMessage)).Methods(coap.GET)
+func serve(svc Service, conn *net.UDPConn, data []byte, addr *net.UDPAddr, rh gocoap.Handler) {
+	msg, err := gocoap.ParseMessage(data)
+	if err != nil {
+		return
+	}
+	var respMsg *gocoap.Message
+	switch msg.Type {
+	case gocoap.Reset:
+		respMsg = &msg
+		cid := mux.Var(&msg, "id")
+		respMsg.Type = gocoap.Acknowledgement
+		publisher, err := Authorize(&msg, respMsg, cid)
+		if err != nil {
+			respMsg.Code = gocoap.Unauthorized
+		} else {
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			err := svc.Unsubscribe(id)
+			if err != nil {
+				respMsg.Code = gocoap.InternalServerError
+			}
+		}
+	default:
+		respMsg = rh.ServeCOAP(conn, addr, &msg)
+	}
+	if respMsg != nil {
+		gocoap.Transmit(conn, addr, *respMsg)
+	}
+}
 
-	r.NotFoundHandler = coap.FuncHandler(notFoundHandler)
+// ListenAndServe binds to the given address and serve requests forever.
+func ListenAndServe(svc Service, mgr manager.ManagerClient, addr string, rh gocoap.Handler) error {
```

Comment:
You are oveloading Dustin's Go server. This is OK for now - probably this way will be easier to bump up his version.

However, on the long-run, I think we'll need to rewrite this server because it is pretty unmaintained and has MIT license.